proposals/82-js-configuration.md

```diff
@@ -0,0 +1,120 @@
+
+Our stack configurations are usually complex, inconsistent and require a lot of prior knowledge of the building blocks of the stack, as well as how these building blocks interact with each other.
+
+With this project, we should improve the configuration experience for the users, as well as reduce its need to power users only. Within the user onboarding scope, configuration across the stack should be simple, complete and establishing a bridge to the concepts behind (visual explanations/illustrations would be rad!). In addition, our set of projects should provide typical configurations that users need for typical scenarios, so that users can start playing without a need for configuration. Finally, within the power users scope, we should land improvements to mitigate all the existing inconsistencies and improve developer experience.
+
+The above scopes can turn this project proposal into two different projects.
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+<!--(bullet list)-->
+
+- Users cannot properly configure our stack autonomously
+- Configuration get users blocked and turn them into inactive users
+
+#### User workflow example
+_How would a developer or user use this new capability?_
+<!--(short paragraph)-->
+
+One of two ways:
+- Inexperienced/First users can use the stack out of the box, without any friction around configurations
+- Power users can create their own configuration for a specific scenario autonomously
+
+#### Impact
+_How would this directly contribute to web3 dev stack product-market fit?_
+
+<!--
+Explain how this addresses known challenges or opportunities.
+What awesome potential impact/outcomes/results will we see if we nail this project?
+-->
+
+#### Leverage
+_How much would nailing this project improve our knowledge and ability to execute future projects?_
+
+<!--
+Explain the opportunity or leverage point for our subsequent velocity/impact (e.g. by speeding up development, enabling more contributors, etc)
+-->
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+<!--Explain why this rating-->
+
+
+## Project definition
+#### Brief plan of attack
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+- We have been cataloging in libp2p all the problems of configuration we see from issues and have some discussions about potential solutions in https://github.com/libp2p/js-libp2p/issues/576. But the scope is much broader than libp2p
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+TODO, we need to define the scope
```

Comment:
Perhaps we can deal with the scoping challenge by limiting it to the libp2p work as round-1 to form stronger opinions about how to do this well and then do subsequent iterations/projects doing that elsewhere. The proposal in https://github.com/libp2p/js-libp2p/issues/576#issuecomment-594590865 looks like a pretty solid suggestion for how to move forward and we could just adopt that as the basis of a plan (with updates to account for newer perspectives).
There's probably some overlap with #71 and some of the work we did in multiformats/ipld (e.g. see https://github.com/multiformats/js-multiformats/blob/master/src/basics.js for a "give me the common basics" version vs the "I'll build my own, thanks"). I imagine there's opportunities here to decouple code paths so that, in configuring your libp2p you're instantiating a set of dependent modules (& packages) that can be easily tree-shaken by a bundler.

Reply:
I think limiting to libp2p initially would be beneficial, it's the most challenging to configure at this point. A possible amendment, although it could be done as a followup, would be to setup deployment to automatically release the various configurations as ready to go libp2p modules, similar to https://github.com/carsonfarmer/libp2p-bundle, which would reduce the need to install any other modules to get up and running. The multiformats/ipld work is a great reference for potential structure here. If we built this out, we could later go through crypto and shake out the needed components for each of these bundles, namely browsers, to reduce size for those configs.

Reply:
I agree with decoupling this into libp2p first as I think that is where most of the pain points are.

However, there are some issues on `ipfs` configuration "conflicts" with libp2p per https://github.com/ipfs/js-ipfs/issues/3590#issuecomment-803509950, where we have multiple ways of configure things and then surprising end results might appear.  This can probably be a maintenance task in ipfs or something that the onboarding team can have a look. What do you think?

Reply:
> This can probably be a maintenance task in ipfs

That or it could eventually become a project. To keep the scope small here, starting with making libp2p easy to construct and then have a default build for Node.js and the browser that can be used on the public network out of the box would be a good starting point. This gets immediate value to new users so they can avoid configuration, and makes it easier for us to roll these changes into js-ipfs and improve its configuration later.